sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

```diff
@@ -130,8 +130,24 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
           if limit < conf.topKSortFallbackThreshold =>
         Some(TakeOrderedAndProjectExec(
           limit, order, projectList, planLater(child)))
+      case Sort(order, true, child) if supportTakeOrdered(child) =>
+        Some(TakeOrderedAndProjectExec(
+          child.maxRows.get.toInt, order, child.output, planLater(child)))
+      case Project(projectList, Sort(order, true, child)) if supportTakeOrdered(child) =>
+        Some(TakeOrderedAndProjectExec(
+          child.maxRows.get.toInt, order, projectList, planLater(child)))
       case _ => None
     }
+
+    private def supportTakeOrdered(plan: LogicalPlan): Boolean = {
+      // Maximum support up to 655360, a larger number of rows does not improve performance.
+      plan.maxRows.exists(_ < math.min(conf.topKSortFallbackThreshold, 655360)) &&
+        // The plan should not contain global sort, to avoid sorting it again.
+        !plan.exists {
+          case s: Sort => s.global
+          case _ => false
+        }
```

Comment:
It is because `RemoveRedundantSorts` can not support remove `TakeOrderedAndProjectExec`. `TakeOrderedAndProjectExec` always keep a single partition: https://github.com/apache/spark/blob/c9ff3603509400fdac50b8f9340d38e1aac91c6f/sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala#L332.